cli/src/commands/gerrit/upload.rs

```diff
@@ -0,0 +1,389 @@
+        }
+
+        // Gerrit change id is 40 chars, jj change id is 32, so we need padding.
+        // To be consistent with `format_gerrit_change_id_trailer``, we pad with
+        // 6a6a6964 (hex of "jjid").
+        let gerrit_change_id = format!("I6a6a6964{}", original_commit.change_id().hex());
+
+        let new_description = format!(
+            "{}{}Change-Id: {}\n",
+            description.trim(),
+            if trailers.is_empty() { "\n\n" } else { "\n" },
+            gerrit_change_id
+        );
+
+        let new_parents = original_commit
+            .parents()
+            .map(|parent| {
+                let p = parent.unwrap();
+                if let Some(rewritten_parent) = old_to_new.get(p.id()) {
+                    rewritten_parent
+                } else {
+                    &p
+                }
+                .id()
+                .clone()
+            })
+            .collect();
+
+        // rewrite the set of parents to point to the commits that were
+        // previously rewritten in toposort order
+        //
+        // TODO FIXME (aseipp): this whole dance with toposorting, calculating
+        // new_parents, and then doing rewrite_commit is roughly equivalent to
+        // what we do in duplicate.rs as well. we should probably refactor this?
+        let new_commit = tx
+            .repo_mut()
+            .rewrite_commit(&original_commit)
+            .set_description(new_description)
+            .set_parents(new_parents)
+            // Set the timestamp back to the timestamp of the original commit.
+            // Otherwise, `jj gerrit upload @ && jj gerrit upload @` will upload
+            // two patchsets with the only difference being the timestamp.
+            .set_committer(original_commit.committer().clone())
+            .set_author(original_commit.author().clone())
+            .write()?;
+
+        old_to_new.insert(original_commit.id().clone(), new_commit.clone());
+    }
+    writeln!(ui.stderr())?;
+
+    let remote_ref = format!("refs/for/{remote_branch}");
+    writeln!(
+        ui.stderr(),
+        "Found {} heads to push to Gerrit (remote '{}'), target branch '{}'",
+        old_heads.len(),
+        remote,
+        remote_branch,
+    )?;
+
+    writeln!(ui.stderr())?;
+
+    // NOTE (aseipp): because we are pushing everything to the same remote ref,
+    // we have to loop and push each commit one at a time, even though
+    // push_updates in theory supports multiple GitRefUpdates at once, because
+    // we obviously can't push multiple heads to the same ref.
+    for head in &old_heads {
+        write!(
+            ui.stderr(),
+            "{}",
+            if upload.dry_run {
+                "Dry-run: Would push "
+            } else {
+                "Pushing "
+            }
+        )?;
+        // We have to write the old commit here, because the until we finish
+        // the transaction (which we don't), the new commit is labelled as
```

Comment:
hmm, why don't we finish the transaction? it seems to me like we should

Reply:
It is how gerrit works for every git user, but:
* We *can* avoid the problem. It's unavoidable in git precisely because git doesn't have another kind of change-id
* Because of how easy jj makes split & squash, jj users split and squash far more frequently than git users, and are thus far more likely to run into this issue.
* As you say, it is more confusing when jj has its own change-id

I've now added documentation here locally. Re user-visible documentation, I've avoided mentioning implementation details, but in that follow-up documentation PR, it's discussed from a "how to use it" perspective.

Reply:
> I think we should at least mention to the user that what we push to Gerrit is a temporary commit with the Change-Id trailer added.

Good point. Done that in my local copy.

> Do you think we should have a jj gerrit download that strips the the Change-Id trailer too? 

I think a `jj gerrit download` (or `jj gerrit patch`) makes a lot of sense, but I don't think it would strip the `Change-Id`. Otherwise `jj gerrit download && jj gerrit upload` would make a new commit (since the jj change-Id differed).

Reply:
Makes sense.

Btw, do you think it would be useful to have `jj gerrit upload` print a message whenever there's a Change-Id trailer with a different value than what would be generated from jj's change id?

Reply:
I don't think so. If it never finishes the transaction, then it has to be manually added by the user. so it would be very intentional.